### ISSUE TITLE
String (slice) add assign String error

### ISSUE THREAD

**Issue by SWW13** [rune-rs/rune#436]
This is likely just a wrong error message, I assume there is a string slice type which can't be extended.

```rust
pub fn main() {
  // bad argument #0: expected `String`, but found `String`
  let str = "Test";
  str += format!("{}", 42);

  // works
  let str = "Test";
  str = str + format!("{}", 42);
  str += format!("{}", 42);
}
```

```rust
error: internal vm error
  ┌─ test.rn:4:3
  │
4 │   str += format!("{}", 42);
  │   ^^^^^^^^^^^^^^^^^^^^^^^^ in this expression
  │
  = bad argument #0: expected `String`, but found `String`

backtrace:
test.rn:4:3:   str += format!("{}", 42);
```

**Comment by udoprog** [comment]
After a bit of investigation:

The kind of relocation that needs to be performed to support this isn't supported with our current addressing mode, because once an operation like `<lhs> += <rhs>` is encountered the first thing the compiler does is clone the values onto two slots on the stack `[.., <lhs>, <rhs>]`. Due to the internal layout of static strings (which is `Arc<StaticString>` ) there is no way to promote these to regular strings on a per-operation basis.

A new addressing mode is investigated which supports referencing variable slots directly for calls like these instead of cloning values onto the stack. That would allow the operation to relocate the static string to be dynamic instead in preparation of performing the mutable call.

The intermediate solution is to explicitly build a dynamic string instead, like the suggested operation above achieves, or by using `let str = String::from_str("Test")`.

**Comment by SWW13** [comment]
Fixed in `0.13.1`
